### PR TITLE
ENT-4896: Added a task to fetch metadata of missing courses, this task will execute as inside update content metadata management command.

### PR DESCRIPTION
**Pull request by saleem-latif**
## Description

The content metadata in enterprise-catalog needs to be augmented for courses within the programs. While traversing courses for a Program, add an alert log if some course is part of the Program and not in the Enterprise catalog.

The endpoints to be traversed are Discovery’s search/all, api/vi/program and api/v1/courseThis work is very similar to existing code which uses the search/all and /api/vi/course endpoints

Description of changes made

## Ticket Link
https://openedx.atlassian.net/browse/ENT-4896

## Post-review

Squash commits into discrete sets of changes

### CODE REVIEW

File: enterprise_catalog/apps/api/tasks.py [hunk 1]
```diff
@@ -628,3 +628,47 @@ def update_catalog_metadata_task(self, catalog_query_id, force=False):  # pylint
...
+    Arguments:
+        force (bool): If true, forces execution of task and ignores time since last run.
+    """
+
+    program_metadata_list = ContentMetadata.objects.filter(content_type=PROGRAM).values_list('json_metadata', flat=True)
+    course_keys = set()
+    for program_metadata in program_metadata_list:
+        if program_metadata is not None:
+            program_metadata.get('courses')
+        course_keys.update([item.get('key') for item in program_metadata.get('courses', [])])
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/api/tasks.py]: if `program_metadata` is `None` then `program_metadata.get('courses', [])` will raise an exception.

File: enterprise_catalog/apps/api/tasks.py [hunk 2]
```diff
@@ -628,3 +628,47 @@ def update_catalog_metadata_task(self, catalog_query_id, force=False):  # pylint
...
+
+    # Check which courses do not have content metadata.
+    present_course_keys = ContentMetadata.objects.\
+        filter(content_type=COURSE, content_key__in=course_keys)\
+        .values_list('content_key', flat=True)
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/api/tasks.py]: how about
```python
present_course_keys = ContentMetadata.objects.filter(
    content_type=COURSE, 
    content_key__in=course_keys
).values_list(
    'content_key', 
    flat=True
)
```

File: enterprise_catalog/apps/api/tasks.py [hunk 3]
```diff
@@ -628,3 +628,47 @@ def update_catalog_metadata_task(self, catalog_query_id, force=False):  # pylint
...
+def fetch_missing_course_metadata_task(self):  # pylint: disable=unused-argument
+    """
+    Creates a CatalogQuery for all the courses that do not have ContentMetadata instance.
+
+    After creating the catalog query it calls update_contentmetadata_from_discovery to update the metadata for these
+    courses. Course metadata is only missing for program courses so the initial query only looks for course metadata
+    that are embedded inside a program.
+
+    Arguments:
+        force (bool): If true, forces execution of task and ignores time since last run.
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/api/tasks.py]: there is no `force` argument for `fetch_missing_course_metadata_task`?

**Comment by saleem-latif** [comment]
@muhammad-ammar I have incorporated your feedback changes, please take another look.

**Review by muhammad-ammar — APPROVED** [review]
LGTM

**Review by sameenfatima78 — APPROVED** [review]

**Pull request merged**